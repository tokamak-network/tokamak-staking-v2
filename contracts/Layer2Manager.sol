// SPDX-License-Identifier: AGPL-3.0-or-later
pragma solidity ^0.8.4;

import "./storages/Layer2ManagerStorage.sol";
import "./proxy/BaseProxyStorage.sol";
import "./common/AccessibleCommon.sol";
import "./libraries/SafeERC20.sol";
import "./libraries/Layer2.sol";
import "./libraries/LibOptimism.sol";

import "hardhat/console.sol";

interface AddressManagerI {
    function getAddress(string memory _name) external view returns (address);
}

interface SeigManagerV2I {
    function claim(address to, uint256 amount) external;
}

interface StakingLayer2I {
    function balanceOfLton(uint32 layerKey, address account) external view returns (uint256 amount);
}

interface SequencerI {
    function create(uint32 _index, bytes memory _layerInfo) external returns (bool);
    function getTvl(uint32 _index) external view returns (uint256);
    function getTvl(address l1Bridge, address l2ton) external view returns (uint256 amount);
    function sequencer(uint32 _index) external view returns (address);
    function sequencer(address addressManager) external view returns (address);
    function layerInfo(uint32 _index) external view returns (bytes memory);
    function getLayerInfo(uint32 _index) external view returns (LibOptimism.Info memory _layerInfo );
}

interface CandidateI {
    function create (uint32 _candidateIndex, bytes memory _data) external returns (bool);
    function layerInfo (uint32 _index) external view returns (bytes memory);
}

contract  Layer2Manager is AccessibleCommon, BaseProxyStorage, Layer2ManagerStorage {
    /* ========== DEPENDENCIES ========== */
    using SafeERC20 for IERC20;

    event Claimed(uint32 _index, address _sequencer, uint256 amount);
    event CreatedOptimismSequencer(uint32 _index, address _sequencer, bytes32 _name, address addressManager, address l1Messenger, address l1Bridge, address l2ton, uint256 depositAmount);
    event CreatedCandidate(uint32 _index, address _operator, bytes32 _name, uint32 _sequencerIndex, uint16 _commission);
    event Distributed(uint256 _totalSeigs, uint256 _distributedAmount);
    event IncreasedSecurityDeposit(uint32 _index, address caller, uint256 amount);
    event DecreasedSecurityDeposit(uint32 _index, address _sequencer, uint256 amount);

    /* ========== CONSTRUCTOR ========== */
    constructor() {
    }

    /* ========== onlyOwner ========== */
    function setMaxLayer2Count(uint256 _maxLayer2Count)
        external nonZero(_maxLayer2Count)
        onlyOwner
    {
        require(maxLayer2Count != _maxLayer2Count, "same");
        maxLayer2Count = _maxLayer2Count;
    }

    function setMinimumDepositForSequencer(uint256 _minimumDepositForSequencer)
        external
        onlyOwner
    {
        require(minimumDepositForSequencer != _minimumDepositForSequencer, "same");
        minimumDepositForSequencer = _minimumDepositForSequencer;
    }

    function setRatioSecurityDepositOfTvl(uint16 _ratioSecurityDepositOfTvl)
        external
        onlyOwner
    {
        require(ratioSecurityDepositOfTvl != _ratioSecurityDepositOfTvl, "same");
        ratioSecurityDepositOfTvl = _ratioSecurityDepositOfTvl;
    }

    function setMinimumDepositForCandidate(uint256 _minimumDepositForCandidate)
        external
        onlyOwner
    {
        require(minimumDepositForCandidate != _minimumDepositForCandidate, "same");
        minimumDepositForCandidate = _minimumDepositForCandidate;
    }

    function setDelayBlocksForWithdraw(uint256 _delayBlocksForWithdraw)
        external
        onlyOwner
    {
        require(delayBlocksForWithdraw != _delayBlocksForWithdraw, "same");
        delayBlocksForWithdraw = _delayBlocksForWithdraw;
    }

    /* ========== only SeigManagerV2 ========== */
    function addSeigs(uint256 amount) external returns (bool)
    {
        require(msg.sender == seigManagerV2, "caller is not SeigManagerV2");
        if (amount > 0) totalSeigs += amount;
        return true;
    }

    /* ========== Sequncer can execute ========== */

    function createOptimismSequencer(
        bytes32 _name,
        address addressManager,
        address l1Messenger,
        address l1Bridge,
        address l2ton
    )
        external ifFree
    {
        require(msg.sender == AddressManagerI(addressManager).getAddress('OVM_Sequencer'), 'NOT Sequencer');

        require(indexSequencers < maxLayer2Count, 'exceeded maxLayer2Count');

        require(
            addressManager != address(0) &&
            l1Messenger != address(0) &&
            l1Bridge != address(0) &&
            l2ton != address(0), "zero address"
        );

        bytes32 _key = LibOptimism.getKey(addressManager, l1Messenger, l1Bridge, l2ton);
        require(!layerKeys[_key], 'already created');
        layerKeys[_key] = true;

        uint32 _index = ++indexSequencers;

        uint256 depositAmount =  minimumSecurityDepositAmount(l1Bridge, l2ton);
        totalSecurityDeposit += depositAmount;
        optimismSequencerIndexes.push(_index);
        Layer2.Layer2Holdings storage holding = holdings[_index];
        holding.securityDeposit = depositAmount;
        optimismSequencerNames[_index] = _name;

        require(
            SequencerI(optimismSequencer).create(_index, abi.encodePacked(addressManager, l1Messenger, l1Bridge, l2ton)),
            "Fail createOptimismSequencer"
        );

        ton.safeTransferFrom(msg.sender, address(this), depositAmount);

        emit CreatedOptimismSequencer(
            _index, msg.sender, _name, addressManager, l1Messenger, l1Bridge, l2ton, depositAmount);
    }

    function createCandidate(
        uint32 _sequencerIndex,
        bytes32 _name,
        uint16 _commission
    )   external nonZeroUint32(_sequencerIndex) ifFree
    {
        require(_sequencerIndex <= indexSequencers, "wrong index");

        bytes32 _key = LibOperator.getKey(msg.sender, _sequencerIndex);
        require(!layerKeys[_key], 'already created');
        layerKeys[_key] = true;

        uint32 _index = ++indexCandidates;

        candidatesIndexes.push(_index);
        candidateNames[_index] = _name;

        ton.safeTransferFrom(msg.sender, address(this), minimumDepositForCandidate);

        if (ton.allowance(address(this), candidate) < minimumDepositForCandidate) {
            ton.approve(candidate, ton.totalSupply());
        }

        require(
            CandidateI(candidate).create(
                _index,
                abi.encodePacked(msg.sender, _sequencerIndex, _commission, minimumDepositForCandidate)),
            "Fail createCandidate"
        );

        emit CreatedCandidate(_index, msg.sender, _name, _sequencerIndex, _commission);
    }

    function decreaseSecurityDeposit(uint32 _sequencerIndex, uint256 amount)
        external ifFree nonZeroUint32(_sequencerIndex)
    {
        require(_sequencerIndex <= indexSequencers, "wrong index");

        LibOptimism.Info memory _layerInfo = SequencerI(optimismSequencer).getLayerInfo(_sequencerIndex);
        address _sequencer = SequencerI(optimismSequencer).sequencer(_layerInfo.addressManager);
        require(_sequencer != address(0) && _sequencer == msg.sender, "sequencer is zero or not caller." );

        uint256 minDepositAmount = minimumSecurityDepositAmount(_layerInfo.l1Bridge, _layerInfo.l2ton);
        Layer2.Layer2Holdings storage holding = holdings[_sequencerIndex];
        require(amount + minDepositAmount <= holding.securityDeposit, "insufficient deposit");

        holding.securityDeposit -= amount;
        totalSecurityDeposit -= amount;
        ton.safeTransfer(msg.sender, amount);

        emit DecreasedSecurityDeposit(_sequencerIndex, msg.sender, amount);
    }

    /* ========== Anyone can execute ========== */

    function increaseSecurityDeposit(uint32 _sequencerIndex, uint256 amount)
        external ifFree nonZeroUint32(_sequencerIndex)
    {
        require(_sequencerIndex <= indexSequencers, "wrong index");
        ton.safeTransferFrom(msg.sender, address(this), amount);

        Layer2.Layer2Holdings storage holding = holdings[_sequencerIndex];
        holding.securityDeposit += amount;
        totalSecurityDeposit += amount;
        emit IncreasedSecurityDeposit(_sequencerIndex, msg.sender, amount);
    }

    function distribute() external {
        require (totalSeigs != 0, 'no distributable amount');
        uint256 len = optimismSequencerIndexes.length;
        uint256 sum = 0;

        uint256[] memory amountLayer = new uint256[](len);
        for(uint256 i = 0; i < len; i++){
            uint32 _layerIndex = optimismSequencerIndexes[i];
            Layer2.Layer2Holdings memory holding = holdings[_layerIndex];

            if (holding.securityDeposit >= minimumDepositForSequencer ) {
                amountLayer[i] += depositsOf(_layerIndex);
            }
            amountLayer[i] += holding.securityDeposit;
            sum += amountLayer[i];
        }
        uint256 amount1 = 0;
        if (sum > 0) {
            for(uint256 i = 0; i < len; i++){
                uint32 _layerIndex = optimismSequencerIndexes[i];
                Layer2.Layer2Holdings storage holding = holdings[_layerIndex];
                if (amountLayer[i] > 0 ) {
                    uint256 amount = totalSeigs * amountLayer[i] / sum;
                    holding.seigs += amount;
                    amount1 += amount;
                }
            }
            if (amount1 > 0)  totalSeigs -= amount1;
        }

        emit Distributed(totalSeigs, amount1);
    }

    function claim(uint32 _layerIndex) external {
        uint256 amount = holdings[_layerIndex].seigs;
        require(amount != 0, 'no amount to claim');
        address sequencer_ = sequencer(_layerIndex);
        require(sequencer_ != address(0), 'zero sequencer');
        holdings[_layerIndex].seigs = 0;
        SeigManagerV2I(seigManagerV2).claim(sequencer_, amount);
        emit Claimed(_layerIndex, sequencer_, amount);
    }

    /* ========== VIEW ========== */
    function minimumSecurityDepositAmount(address l1Bridge, address l2ton) public view returns (uint256 amount) {
        if (ratioSecurityDepositOfTvl == 0) amount = minimumDepositForSequencer;
        else {
            amount = Math.max(
                SequencerI(optimismSequencer).getTvl(l1Bridge, l2ton) * ratioSecurityDepositOfTvl / 10000,
                minimumDepositForSequencer);
        }
    }

    function balanceOfLton(address account) public view returns (uint256 amount) {
        uint256 len = optimismSequencerIndexes.length;
        for(uint256 i = 0; i < len; i++){
            amount += StakingLayer2I(optimismSequencer).balanceOfLton(optimismSequencerIndexes[i], account);
        }
    }

    function curTotalLayer2Deposits() public view returns (uint256 amount) {
        uint256 len = optimismSequencerIndexes.length;
        for(uint256 i = 0; i < len; i++){
            amount += SequencerI(optimismSequencer).getTvl(optimismSequencerIndexes[i]);
        }
    }

    function sequencer(uint32 _layerIndex) public view returns (address sequencer_) {
        if (_layerIndex <= indexSequencers ){
            sequencer_ = SequencerI(optimismSequencer).sequencer(_layerIndex);
        }
    }

    function existedLayer2Index(uint32 _index) external view returns (bool exist_) {
        if (_index <= indexSequencers) exist_ = true;
    }

    function existedCandidateIndex(uint32 _index) external view returns (bool exist_) {
        if (_index <= indexCandidates) exist_ = true;
    }

    function curTotalAmountsLayer2() external view returns (uint256 amount) {
        amount = curTotalLayer2Deposits() + totalSecurityDeposit;
    }

    function totalLayers() external view returns (uint256 total) {
        total = optimismSequencerIndexes.length;
    }

    function totalCandidates() external view returns (uint256 total) {
        total = candidatesIndexes.length;
    }

    function curlayer2DepositsOf(uint32 _layerIndex) external view returns (uint256 amount) {
        amount = depositsOf(_layerIndex);
    }

    function depositsOf(uint32 _layerIndex) public view returns (uint256 amount) {
        try
            SequencerI(optimismSequencer).getTvl(_layerIndex) returns (uint256 a) {
                amount = a;
        } catch (bytes memory ) {
            amount = 0;
        }
    }

    function getAllLayers()
        external view
        returns (
            bytes32[] memory optimismSequencerNames_,
            uint32[] memory optimismSequencerIndexes_,
            Layer2.Layer2Holdings[] memory holdings_,
            bytes[] memory infos_
            )
    {
        uint256 len = optimismSequencerIndexes.length;

        optimismSequencerNames_ = new bytes32[](len);
        optimismSequencerIndexes_ = optimismSequencerIndexes;
        holdings_ = new Layer2.Layer2Holdings[](len);
        infos_ = new bytes[](len);
        for (uint256 i = 0; i < len ; i++){
            optimismSequencerNames_[i] = optimismSequencerNames[optimismSequencerIndexes[i]];
            holdings_[i] = holdings[optimismSequencerIndexes[i]];
            infos_[i] = SequencerI(optimismSequencer).layerInfo(optimismSequencerIndexes[i]);
        }
    }

    function getAllCandidates()
        external view
        returns (
            bytes32[] memory candidateNames_,
            uint32[] memory candidateNamesIndexes_,
            bytes[] memory infos_
            )
    {
        uint256 len = candidatesIndexes.length;

        candidateNames_ = new bytes32[](len);
        candidateNamesIndexes_ = candidatesIndexes;
        infos_ = new bytes[](len);
        for (uint256 i = 0; i < len ; i++){
            candidateNames_[i] = candidateNames[candidatesIndexes[i]];
            infos_[i] = CandidateI(candidate).layerInfo(candidatesIndexes[i]);
        }
    }

    function layerHoldings(uint32 layerKey_)
        external view
        returns (Layer2.Layer2Holdings memory)
    {
        return holdings[layerKey_];
    }

    /* ========== internal ========== */
}
