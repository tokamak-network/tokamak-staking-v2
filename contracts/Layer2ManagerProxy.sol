// SPDX-License-Identifier: AGPL-3.0-or-later
pragma solidity ^0.8.4;

import "./Layer2ManagerStorage.sol";
import "./proxy/BaseProxy.sol";

contract Layer2ManagerProxy is BaseProxy, Layer2ManagerStorage {

    function initialize(
        address _ton,
        address _seigManagerV2,
        address _stakingLayer2,
        uint256 _minimumDepositForSequencer,
        uint32 _delayBlocksForWithdraw
    )
        external onlyOwner
        nonZeroAddress(_ton)
        nonZeroAddress(_seigManagerV2)
        nonZeroAddress(_stakingLayer2)
        nonZero(_minimumDepositForSequencer)
        nonZero(_delayBlocksForWithdraw)
    {
        require(address(ton) == address(0), "already initialize");

        ton = IERC20(_ton);
        seigManagerV2 = _seigManagerV2;
        stakingLayer2 = _stakingLayer2;

        minimumDepositForSequencer = _minimumDepositForSequencer;
        delayBlocksForWithdraw = _delayBlocksForWithdraw;

        maxLayer2Count = 5;
    }
}