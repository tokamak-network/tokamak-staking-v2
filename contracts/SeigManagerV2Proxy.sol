// SPDX-License-Identifier: AGPL-3.0-or-later
pragma solidity ^0.8.4;

import "./SeigManagerV2Storage.sol";
import "./proxy/BaseProxy.sol";

contract SeigManagerV2Proxy is BaseProxy, SeigManagerV2Storage {

    /* ========== onlyOwner ========== */

    function initialize(
        address _ton,
        address _wton,
        address _tot,
        address _seigManagerV1,
        address _layer2Manager,
        address _stakingLayer2,
        uint256 _seigPerBlock,
        uint32 _minimumBlocksForUpdateSeig
    )
        external onlyOwner
    {
        require(
            _ton != address(0) &&
            _wton != address(0) &&
            _tot != address(0) &&
            _seigManagerV1 != address(0) &&
            _stakingLayer2 != address(0) &&
            _layer2Manager != address(0) &&
            _seigPerBlock != 0
            , "P1");

        require(address(ton) == address(0), "already initialize");

        seigManagerV1 = _seigManagerV1;
        ton = IERC20(_ton);
        wton = _wton;
        tot = _tot;
        layer2Manager =_layer2Manager;
        stakingLayer2 = _stakingLayer2;
        seigPerBlock = _seigPerBlock;
        minimumBlocksForUpdateSeig = _minimumBlocksForUpdateSeig;

        indexSton = 1 ether;
        indexLton = 1 ether;
    }
}