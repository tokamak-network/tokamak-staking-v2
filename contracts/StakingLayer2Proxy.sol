// SPDX-License-Identifier: AGPL-3.0-or-later
pragma solidity ^0.8.4;

import "./StakingLayer2Storage.sol";
import "./proxy/BaseProxy.sol";

contract StakingLayer2Proxy is BaseProxy, StakingLayer2Storage {

    function initialize(
        address _ton,
        address _seigManagerV2,
        address _layer2Manager
    )
        external onlyOwner
        nonZeroAddress(_ton)
        nonZeroAddress(_seigManagerV2)
        nonZeroAddress(_layer2Manager)
    {
        require(address(ton) == address(0), "already initialize");

        seigManagerV2 = _seigManagerV2;
        ton = _ton;
        layer2Manager =_layer2Manager;
    }
}