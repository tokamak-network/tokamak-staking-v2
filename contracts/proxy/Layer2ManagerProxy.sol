// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

import "../storages/Layer2ManagerStorage.sol";
import "../proxy/BaseProxy.sol";

contract Layer2ManagerProxy is BaseProxy, Layer2ManagerStorage {

    function initialize(
        address _ton,
        address _seigManagerV2,
        address _optimismL2Operator,
        address _candidate,
        uint256 _minimumDepositForL2Operator,
        uint256 _minimumDepositForCandidate,
        uint32 _delayBlocksForWithdraw
    )
        external onlyOwner
        nonZeroAddress(_ton)
        nonZeroAddress(_seigManagerV2)
        nonZeroAddress(_optimismL2Operator)
        nonZeroAddress(_candidate)
        nonZero(_delayBlocksForWithdraw)
    {
        require(address(ton) == address(0), "already initialize");

        ton = IERC20(_ton);
        seigManagerV2 = _seigManagerV2;
        optimismL2Operator = _optimismL2Operator;
        candidate = _candidate;

        minimumDepositForL2Operator = _minimumDepositForL2Operator;
        minimumDepositForCandidate = _minimumDepositForCandidate;
        delayBlocksForWithdraw = _delayBlocksForWithdraw;

        maxLayer2Count = 5;
    }
}