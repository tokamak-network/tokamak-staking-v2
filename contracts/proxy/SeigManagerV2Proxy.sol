// SPDX-License-Identifier: AGPL-3.0-or-later
pragma solidity ^0.8.4;

import "../storages/SeigManagerV2Storage.sol";
import "../proxy/BaseProxy.sol";

contract SeigManagerV2Proxy is BaseProxy, SeigManagerV2Storage {

    /* ========== onlyOwner ========== */

    function initialize(
        address _ton,
        address _wton,
        address _tot,
        address _seigManagerV1,
        address _layer2Manager,
        address _optimismSequencer,
        address _candidate,
        uint256 _seigPerBlock,
        uint32 _minimumBlocksForUpdateSeig
    )
        external onlyOwner
    {
        require(
            _ton != address(0) &&
            _wton != address(0) &&
            _tot != address(0) &&
            _seigManagerV1 != address(0) &&
            _optimismSequencer != address(0) &&
            _candidate != address(0) &&
            _layer2Manager != address(0) &&
            _seigPerBlock != 0
            , "P1");

        require(address(ton) == address(0), "already initialize");

        seigManagerV1 = _seigManagerV1;
        ton = IERC20(_ton);
        wton = _wton;
        tot = _tot;
        layer2Manager =_layer2Manager;
        optimismSequencer = _optimismSequencer;
        candidate = _candidate;

        seigPerBlock = _seigPerBlock;
        minimumBlocksForUpdateSeig = _minimumBlocksForUpdateSeig;

        indexLton = 1 ether;
    }
}