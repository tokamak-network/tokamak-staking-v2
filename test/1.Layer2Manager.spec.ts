import { expect } from './shared/expect'
import { ethers, network } from 'hardhat'

import { Signer } from 'ethers'
import {stakingV2Fixtures} from './shared/fixtures'
import {TonStakingV2Fixture } from './shared/fixtureInterfaces'
/*
import { SeigManagerV2Proxy } from '../typechain-types/contracts/SeigManagerV2Proxy'
import { SeigManagerV2 } from '../typechain-types/contracts/SeigManagerV2.sol'
import { Layer2ManagerProxy } from '../typechain-types/contracts/Layer2ManagerProxy'
import { Layer2Manager } from '../typechain-types/contracts/Layer2Manager.sol'
import { StakingLayer2Proxy } from '../typechain-types/contracts/StakingLayer2Proxy'
import { StakingLayer2 } from '../typechain-types/contracts/StakingLayer2.sol'
import { TON } from '../typechain-types/contracts/test/TON.sol'
import { Lib_AddressManager } from '../typechain-types/contracts/test/Lib_AddressManager'
import { MockL1Messenger } from '../typechain-types/contracts/test/MockL1Messenger'
import { MockL2Messenger } from '../typechain-types/contracts/test/MockL2Messenger'
import { MockL1Bridge } from '../typechain-types/contracts/test/MockL1Bridge'
import { MockL2Bridge } from '../typechain-types/contracts/test/MockL2Bridge'
import { TestERC20 } from '../typechain-types/contracts/test/TestERC20'

interface TonStakingV2Fixture  {
    seigManagerV2Proxy: SeigManagerV2Proxy
    seigManagerV2: SeigManagerV2
    layer2ManagerProxy: Layer2ManagerProxy
    layer2Manager: Layer2Manager
    stakingLayer2Proxy: StakingLayer2Proxy
    stakingLayer2: StakingLayer2
    ton: TON,
    deployer: Signer,
    addr1: Signer,
    addr2: Signer,
    tonAdmin: Signer,
    addressManager: Lib_AddressManager,
    l1Messenger: MockL1Messenger,
    l2Messenger: MockL2Messenger,
    l1Bridge: MockL1Bridge,
    l2Bridge: MockL2Bridge,
    l2ton: TestERC20
}
*/
describe('Layer2Manager', () => {
    let deployer: Signer, addr1: Signer, sequencer1:Signer

    let deployed: TonStakingV2Fixture

    // mainnet
    let seigManagerInfo = {
        ton: "0x2be5e8c109e2197D077D13A82dAead6a9b3433C5",
        wton: "0xc4A11aaf6ea915Ed7Ac194161d2fC9384F15bff2",
        tot: "0x6FC20Ca22E67aAb397Adb977F092245525f7AeEf",
        seigManagerV1: "0x710936500aC59e8551331871Cbad3D33d5e0D909",
        layer2Manager: "",
        seigPerBlock: ethers.BigNumber.from("3920000000000000000"),
        minimumBlocksForUpdateSeig: 300,
    }

    let layer2ManagerInfo = {
        minimumDepositForSequencer: ethers.utils.parseEther("100"),
        delayBlocksForWithdraw: 300,
    }

    before('create fixture loader', async () => {
        // deployed = await loadFixture(stakingV2Fixtures)

        deployed = await stakingV2Fixtures()
        deployer = deployed.deployer;
        addr1 = deployed.addr1;
        sequencer1 = deployed.sequencer1;
    })

    describe('# initialize', () => {

        it('initialize can not be executed by not owner', async () => {
            await expect(
                deployed.layer2ManagerProxy.connect(addr1).initialize(
                    seigManagerInfo.ton,
                    deployed.seigManagerV2Proxy.address,
                    deployed.stakingLayer2Proxy.address,
                    layer2ManagerInfo.minimumDepositForSequencer,
                    layer2ManagerInfo.delayBlocksForWithdraw
                )
                ).to.be.revertedWith("Accessible: Caller is not an admin")
        })

        it('initialize can be executed by only owner', async () => {
            await deployed.layer2ManagerProxy.connect(deployer).initialize(
                    seigManagerInfo.ton,
                    deployed.seigManagerV2Proxy.address,
                    deployed.stakingLayer2Proxy.address,
                    layer2ManagerInfo.minimumDepositForSequencer,
                    layer2ManagerInfo.delayBlocksForWithdraw
                );

            expect(await deployed.layer2ManagerProxy.ton()).to.eq(seigManagerInfo.ton)
            expect(await deployed.layer2ManagerProxy.seigManagerV2()).to.eq(deployed.seigManagerV2Proxy.address)
            expect(await deployed.layer2ManagerProxy.stakingLayer2()).to.eq(deployed.stakingLayer2Proxy.address)
            expect(await deployed.layer2ManagerProxy.minimumDepositForSequencer()).to.eq(layer2ManagerInfo.minimumDepositForSequencer)
            expect(await deployed.layer2ManagerProxy.delayBlocksForWithdraw()).to.eq(layer2ManagerInfo.delayBlocksForWithdraw)
        })

        it('can execute only once.', async () => {
            await expect(
                deployed.layer2ManagerProxy.connect(deployer).initialize(
                    seigManagerInfo.ton,
                    deployed.seigManagerV2Proxy.address,
                    deployed.stakingLayer2Proxy.address,
                    layer2ManagerInfo.minimumDepositForSequencer,
                    layer2ManagerInfo.delayBlocksForWithdraw
                )
                ).to.be.revertedWith("already initialize")
        })
    });

    describe('# setMaxLayer2Count', () => {

        it('setMaxLayer2Count can not be executed by not owner', async () => {
            const maxLayer2Count = ethers.BigNumber.from("2");
            await expect(
                deployed.layer2Manager.connect(addr1).setMaxLayer2Count(maxLayer2Count)
                ).to.be.revertedWith("Accessible: Caller is not an admin")
        })

        it('setMaxLayer2Count can be executed by only owner ', async () => {
            const maxLayer2Count = ethers.BigNumber.from("2");
            await deployed.layer2Manager.connect(deployer).setMaxLayer2Count(maxLayer2Count)
            expect(await deployed.layer2Manager.maxLayer2Count()).to.eq(maxLayer2Count)

            await deployed.layer2Manager.connect(deployer).setMaxLayer2Count(ethers.constants.One)
            expect(await deployed.layer2Manager.maxLayer2Count()).to.eq(ethers.constants.One)
        })

        it('cannot be changed to the same value', async () => {
            const maxLayer2Count = ethers.constants.One
            await expect(
                deployed.layer2Manager.connect(deployer).setMaxLayer2Count(maxLayer2Count)
                ).to.be.revertedWith("same")
        })
    });

    describe('# setMinimumDepositForSequencer', () => {

        it('setMinimumDepositForSequencer can not be executed by not owner', async () => {
            const minimumDepositForSequencer = ethers.utils.parseEther("200");
            await expect(
                deployed.layer2Manager.connect(addr1).setMinimumDepositForSequencer(minimumDepositForSequencer)
                ).to.be.revertedWith("Accessible: Caller is not an admin")
        })

        it('setMinimumDepositForSequencer can be executed by only owner ', async () => {
            const minimumDepositForSequencer = ethers.utils.parseEther("200");
            await deployed.layer2Manager.connect(deployer).setMinimumDepositForSequencer(minimumDepositForSequencer)
            expect(await deployed.layer2Manager.minimumDepositForSequencer()).to.eq(minimumDepositForSequencer)

            await deployed.layer2Manager.connect(deployer).setMinimumDepositForSequencer(layer2ManagerInfo.minimumDepositForSequencer)
            expect(await deployed.layer2Manager.minimumDepositForSequencer()).to.eq(layer2ManagerInfo.minimumDepositForSequencer)
        })

        it('cannot be changed to the same value', async () => {
            await expect(
                deployed.layer2Manager.connect(deployer).setMinimumDepositForSequencer(layer2ManagerInfo.minimumDepositForSequencer)
                ).to.be.revertedWith("same")
        })
    });

    describe('# setDelayBlocksForWithdraw', () => {

        it('setDelayBlocksForWithdraw can not be executed by not owner', async () => {
            const delayBlocksForWithdraw = ethers.BigNumber.from("100");
            await expect(
                deployed.layer2Manager.connect(addr1).setDelayBlocksForWithdraw(delayBlocksForWithdraw)
                ).to.be.revertedWith("Accessible: Caller is not an admin")
        })

        it('setDelayBlocksForWithdraw can be executed by only owner ', async () => {
            const delayBlocksForWithdraw = ethers.BigNumber.from("100");
            await deployed.layer2Manager.connect(deployer).setDelayBlocksForWithdraw(delayBlocksForWithdraw)
            expect(await deployed.layer2Manager.delayBlocksForWithdraw()).to.eq(delayBlocksForWithdraw)


            await deployed.layer2Manager.connect(deployer).setDelayBlocksForWithdraw(layer2ManagerInfo.delayBlocksForWithdraw)
            expect(await deployed.layer2Manager.delayBlocksForWithdraw()).to.eq(layer2ManagerInfo.delayBlocksForWithdraw)
        })

        it('cannot be changed to the same value', async () => {
            await expect(
                deployed.layer2Manager.connect(deployer).setDelayBlocksForWithdraw(layer2ManagerInfo.delayBlocksForWithdraw)
                ).to.be.revertedWith("same")
        })
    });

    describe('#create', () => {

        it('Cannot be created unless the caller is the layer\'s sequencer.', async () => {
            expect(await deployed.addressManager.getAddress("OVM_Sequencer")).to.not.eq(addr1.address)
            await expect(
                deployed.layer2Manager.connect(addr1).create(
                    deployed.addressManager.address,
                    deployed.l1Messenger.address,
                    deployed.l2Messenger.address,
                    deployed.l1Bridge.address,
                    deployed.l2Bridge.address,
                    deployed.l2ton.address
                )
                ).to.be.revertedWith("NOT Sequencer")
        })

        it('If the minimum security deposit is not provided, it cannot be created.', async () => {

            expect(await deployed.addressManager.getAddress("OVM_Sequencer")).to.eq(sequencer1.address)
            await expect(
                deployed.layer2Manager.connect(sequencer1).create(
                    deployed.addressManager.address,
                    deployed.l1Messenger.address,
                    deployed.l2Messenger.address,
                    deployed.l1Bridge.address,
                    deployed.l2Bridge.address,
                    deployed.l2ton.address
                )).to.be.reverted;
        })

        it('Approve the minimum security deposit and create.', async () => {

            expect(await deployed.addressManager.getAddress("OVM_Sequencer")).to.eq(sequencer1.address)
            let totalSecurityDeposit = await deployed.layer2Manager.totalSecurityDeposit();
            let amount = await deployed.layer2Manager.minimumDepositForSequencer();

            if (amount.gt(await deployed.ton.balanceOf(sequencer1.address)))
                await (await deployed.ton.connect(deployed.tonAdmin).mint(sequencer1.address, amount)).wait();


            if (amount.gte(await deployed.ton.allowance(sequencer1.address, deployed.layer2Manager.address)))
                await (await deployed.ton.connect(sequencer1).approve(deployed.layer2Manager.address, amount)).wait();

            await deployed.layer2Manager.connect(sequencer1).create(
                    deployed.addressManager.address,
                    deployed.l1Messenger.address,
                    deployed.l2Messenger.address,
                    deployed.l1Bridge.address,
                    deployed.l2Bridge.address,
                    deployed.l2ton.address
                );

            expect(await deployed.layer2Manager.totalSecurityDeposit()).to.eq(totalSecurityDeposit.add(amount))
        })

    });

});

