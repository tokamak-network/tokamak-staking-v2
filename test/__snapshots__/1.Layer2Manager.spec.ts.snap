// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer2Manager # initialize initialize can be executed by only owner 1`] = `203524`;

exports[`Layer2Manager # setDelayBlocksForWithdraw setDelayBlocksForWithdraw can be executed by only owner  1`] = `38354`;

exports[`Layer2Manager # setMaxLayer2Count setMaxLayer2Count can be executed by only owner  1`] = `38335`;

exports[`Layer2Manager # setMinimumDepositForSequencer setMinimumDepositForSequencer can be executed by only owner  1`] = `38402`;
